File: XRViewer/LocationManager/LocationManager.m
#import "LocationManager.h"
#import <os/lock.h>
#import "WebARKHeader.h"

@interface LocationManager() <CLLocationManagerDelegate>
{
    os_unfair_lock _lock;
}

@property (nonatomic, strong) CLLocationManager *manager;
@property (nonatomic, strong) CLLocation *currentLocation;
@property (nonatomic, copy) NSDictionary *request;
@property (nonatomic, copy) DidRequestAuth authBlock;
@end

@implementation LocationManager

- (void)dealloc
{
    DDLogDebug(@"LocationManager dealloc");
}

- (instancetype)init
{
    self = [super init];
    
    if (self)
    {
        _lock = OS_UNFAIR_LOCK_INIT;
        
        _manager = [[CLLocationManager alloc] init];
        [_manager setDelegate:self];
    }
    
    return self;
}

- (BOOL)requestAuthorization
{
    if ([CLLocationManager authorizationStatus] == kCLAuthorizationStatusNotDetermined)
    {
        [_manager requestWhenInUseAuthorization];
        return NO;
    }
    
    return YES;
}

- (void)requestAuthorizationWithCompletion:(DidRequestAuth)authBlock
{
    [self setAuthBlock:authBlock];
    
    if ([self requestAuthorization])
    {
        if (authBlock)
        {
            if ([CLLocationManager authorizationStatus] == kCLAuthorizationStatusAuthorizedWhenInUse)
            {
                authBlock(YES);
            }
            else
            {
                authBlock(NO);
            }
        }
    }
}

- (void)startUpdateLocation
{
    if ([CLLocationManager authorizationStatus] != kCLAuthorizationStatusAuthorizedWhenInUse)
    {
        DDLogError(@"Location isn't allowed !");
        return;
    }
    
#warning LOCATION TEMP SOLUTION ( waiting for geofencing )
    [_manager requestLocation];
    
    //[_manager startUpdatingLocation];
}

- (void)stopUpdateLocation
{
    if ([CLLocationManager authorizationStatus] != kCLAuthorizationStatusAuthorizedWhenInUse)
    {
        DDLogError(@"Location isn't allowed !");
        return;
    }
    
#warning LOCATION TEMP SOLUTION ( waiting for geofencing )
    //[_manager stopUpdatingLocation];
}

- (CLLocationCoordinate2D)currentCoordinate
{
    CLLocationCoordinate2D coordinate;
    
    os_unfair_lock_lock(&(_lock));
    coordinate = [_currentLocation coordinate];
    os_unfair_lock_unlock(&(_lock));
    
    return coordinate;
}

- (CLLocationDistance)currentAltitude
{
    CLLocationDistance altitude;
    
    os_unfair_lock_lock(&(_lock));
    altitude = [_currentLocation altitude];
    os_unfair_lock_unlock(&(_lock));
    
    return altitude;
}

- (NSDictionary *)currentCoordinateDict
{
    CLLocationCoordinate2D coord = [self currentCoordinate];
    CLLocationDistance altitude = [self currentAltitude];
    
    return @{WEB_AR_LOCATION_LON_OPTION : @(coord.longitude), WEB_AR_LOCATION_LAT_OPTION : @(coord.latitude), WEB_AR_LOCATION_ALT_OPTION : @(altitude)};
}

- (void)setupForRequest:(NSDictionary *)request
{
    [self setRequest:request];
    
    if ([request[WEB_AR_LOCATION_OPTION] boolValue])
    {
        [self startUpdateLocation];
    }
    else
    {
        [self stopUpdateLocation];
    }
}

- (NSDictionary *)locationData
{
    if ([[self request][WEB_AR_LOCATION_OPTION] boolValue])
    {
        return @{WEB_AR_LOCATION_OPTION : [self currentCoordinateDict]};
    }
    
    return @{};
}

#pragma mark Location Manager

- (void)locationManager:(CLLocationManager *)manager
     didUpdateLocations:(NSArray<CLLocation *> *)locations
{
    os_unfair_lock_lock(&(_lock));
    _currentLocation = [[locations lastObject] copy];
    os_unfair_lock_unlock(&(_lock));
    
    if ([self updateLocation])
    {
        [self updateLocation]([locations lastObject]);
    }
}

- (void)locationManager:(CLLocationManager *)manager
       didFailWithError:(NSError *)error
{
    DDLogError(@"Location error - %@", error);
}

- (void)locationManager:(CLLocationManager *)manager didChangeAuthorizationStatus:(CLAuthorizationStatus)status
{
    DDLogDebug(@"locationManager didChangeAuthorizationStatus - %d", status);
    
    if ([self authBlock])
    {
        switch (status)
        {
            case kCLAuthorizationStatusNotDetermined:
                
                break;
            case kCLAuthorizationStatusRestricted:
            case kCLAuthorizationStatusDenied:
                [self authBlock](NO);
                break;
            case kCLAuthorizationStatusAuthorizedAlways:
            case kCLAuthorizationStatusAuthorizedWhenInUse:
                [self authBlock](YES);
                break;
        }
    }
}

@end

File: XRViewer/LocationManager.swift
import CoreLocation
import Foundation
import os
import XCGLogger

typealias DidUpdateLocation = (CLLocation?) -> Void
typealias DidRequestAuth = (Bool) -> Void

class LocationManager: NSObject, CLLocationManagerDelegate {
    
    private var manager: CLLocationManager
    private var currentLocation: CLLocation?
    private var request: [AnyHashable : Any] = [:]
    private var authBlock: DidRequestAuth?
    var updateLocation: DidUpdateLocation?
    private var lock: os_unfair_lock
    
    override init() {
        lock = os_unfair_lock()
        manager = CLLocationManager()
        
        super.init()
        manager.desiredAccuracy = kCLLocationAccuracyBestForNavigation
        manager.distanceFilter = kCLDistanceFilterNone
        manager.headingFilter = kCLHeadingFilterNone
        manager.pausesLocationUpdatesAutomatically = false
        
        manager.delegate = self
    }
    
    deinit {
        appDelegate().logger.debug("LocationManager dealloc")
    }

    func startUpdateLocation() {
        if CLLocationManager.authorizationStatus() != .authorizedWhenInUse {
            appDelegate().logger.error("Location isn't allowed !")
            return
        }

        manager.startUpdatingLocation()
    }

    func stopUpdateLocation() {
        if CLLocationManager.authorizationStatus() != .authorizedWhenInUse {
            appDelegate().logger.error("Location isn't allowed !")
            return
        }
        manager.stopUpdatingLocation()
    }

    func startUpdateHeading() {
        if CLLocationManager.authorizationStatus() != .authorizedWhenInUse {
            appDelegate().logger.error("Location isn't allowed !")
            return
        }
        
        manager.startUpdatingHeading()
    }
    
    func stopUpdateHeading() {
        if CLLocationManager.authorizationStatus() != .authorizedWhenInUse {
            appDelegate().logger.error("Location isn't allowed !")
            return
        }
        manager.stopUpdatingHeading()
    }

    func currentCoordinate() -> CLLocationCoordinate2D {
        var coordinate = CLLocationCoordinate2D()

        os_unfair_lock_lock(&(lock))
        if let aCoordinate = currentLocation?.coordinate {
            coordinate = aCoordinate
        }
        os_unfair_lock_unlock(&(lock))

        return coordinate
    }

    func currentAltitude() -> CLLocationDistance {
        var altitude: CLLocationDistance

        os_unfair_lock_lock(&(lock))
        altitude = currentLocation?.altitude ?? 0
        os_unfair_lock_unlock(&(lock))

        return altitude
    }

    func setup(forRequest request: [AnyHashable : Any]?) {
        guard let request = request else { return }
        self.request = request

        guard let locationOption = request[WEB_AR_LOCATION_OPTION] as? Bool else { return }
        if locationOption {
            startUpdateLocation()
        } else {
            stopUpdateLocation()
        }
        
        guard let orientationOptions = request[WEB_AR_WORLD_ALIGNMENT] as? Bool else {return}
        if orientationOptions {
            startUpdateHeading()
        } else {
            stopUpdateHeading()
        }
    }

    func locationData() -> [AnyHashable : Any]? {
        guard let locationOption = self.request[WEB_AR_LOCATION_OPTION] as? Bool else { return [:] }
        if locationOption {
            if let aDict = currentCoordinateDict() {
                return [WEB_AR_LOCATION_OPTION: aDict]
            }
            return [:]
        }
        return [:]
    }

    func requestAuthorization(withCompletion authBlock: @escaping DidRequestAuth) {
        self.authBlock = authBlock

        if requestAuthorization() {
            //if authBlock

            if CLLocationManager.authorizationStatus() == .authorizedWhenInUse {
                authBlock(true)
            } else {
                authBlock(false)
            }
        }
    }

    func requestAuthorization() -> Bool {
        if CLLocationManager.authorizationStatus() == .notDetermined {
            manager.requestWhenInUseAuthorization()
            return false
        }

        return true
    }

    func currentCoordinateDict() -> [AnyHashable : Any]? {
        let coord: CLLocationCoordinate2D = currentCoordinate()
        let altitude: CLLocationDistance = currentAltitude()

        return [WEB_AR_LOCATION_LON_OPTION: coord.longitude, WEB_AR_LOCATION_LAT_OPTION: coord.latitude, WEB_AR_LOCATION_ALT_OPTION: altitude]
    }

// MARK: Location Manager

    func locationManager(_ manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        os_unfair_lock_lock(&(lock))
        currentLocation = locations.last
        os_unfair_lock_unlock(&(lock))
    }

    func locationManager(_ manager: CLLocationManager, didUpdateHeading newHeading: CLHeading) {
    }
    
    public func locationManagerShouldDisplayHeadingCalibration(_ manager: CLLocationManager) -> Bool {
        return true
    }
    
    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
        appDelegate().logger.error("Location error - \(error)")
    }

    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        appDelegate().logger.debug("locationManager didChangeAuthorizationStatus - \(status)")

        //if authBlock

        switch status {
            case .notDetermined:
                break
            case .restricted, .denied:
                authBlock?(false)
            case .authorizedAlways, .authorizedWhenInUse:
                authBlock?(true)
            default:
                break
        }
    }
}
